File: android/app/src/main/java/it/curzel/bitscape/analytics/AnalyticsService.kt
package it.curzel.bitscape.analytics

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import com.google.firebase.analytics.FirebaseAnalytics
import android.content.Context
import android.os.Bundle
import it.curzel.bitscape.gamecore.NativeLib

class AnalyticsService(
    private val broker: RuntimeEventsBroker,
    private val nativeLib: NativeLib,
    context: Context
) {

    private val firebaseAnalytics: FirebaseAnalytics = FirebaseAnalytics.getInstance(context)
    private val serviceScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var job: Job? = null

    init {
        serviceScope.launch {
            delay(500)
            bindEvents()
        }
    }

    private fun bindEvents() {
        job = broker.events()
            .onEach { handle(it) }
            .launchIn(serviceScope)
    }

    private fun handle(event: RuntimeEvent) {
        event.toAnalyticsEvent(nativeLib)?.let { send(it) }
    }

    private fun send(event: AnalyticsEvent) {
        firebaseAnalytics.logEvent(event.name, event.params)
    }
}

private data class AnalyticsEvent(
    val name: String,
    val params: Bundle? = null
)

private fun RuntimeEvent.toAnalyticsEvent(nativeLib: NativeLib): AnalyticsEvent? {
    return when (this) {
        is RuntimeEvent.Loading -> AnalyticsEvent(name = "app_loading")
        is RuntimeEvent.Launched -> AnalyticsEvent(name = "app_launched")
        is RuntimeEvent.WillEnterForeground -> AnalyticsEvent(name = "will_enter_foreground")
        is RuntimeEvent.DidEnterBackground -> AnalyticsEvent(name = "did_enter_background")
        is RuntimeEvent.NewGame -> AnalyticsEvent(name = "new_game_started")
        is RuntimeEvent.GameOver -> {
            val params = Bundle().apply {
                putInt("current_world", nativeLib.currentWorldId())
                putInt("ammo_count", nativeLib.numberOfKunaiInInventory())
            }
            AnalyticsEvent(name = "game_over", params = params)
        }
        is RuntimeEvent.WorldTransition -> {
            if (source != 0u) {
                val params = Bundle().apply {
                    putInt("source", source.toInt())
                    putInt("destination", destination.toInt())
                }
                AnalyticsEvent(name = "world_transition", params = params)
            } else {
                null
            }
        }
    }
}

File: ios/SneakBit/Analytics/AnalyticsService.swift
import Combine
import FirebaseAnalytics
import Foundation

class FirebaseAnalyticsService {
    @Inject private var broker: RuntimeEventsBroker
    
    private var disposables = Set<AnyCancellable>()
    
    init() {
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            self.bindEvents()
        }
    }
    
    private func bindEvents() {
        broker.events()
            .sink { self.handle($0) }
            .store(in: &self.disposables)
    }
    
    private func handle(_ event: RuntimeEvent) {
        if let event = event.toAnalyticsEvent() {
            send(event)
        }
    }
    
    private func send(_ event: AnalyticsEvent) {
        Analytics.logEvent(event.name, parameters: event.params)
    }
}

private extension RuntimeEvent {
    func toAnalyticsEvent() -> AnalyticsEvent? {
        switch self {
        case .loading: .init(name: "app_loading")
        case .launched: .init(name: "app_launched")
        case .willEnterForeground: .init(name: "will_enter_foreground")
        case .didEnterBackground: .init(name: "did_enter_background")
        case .newGame: .init(name: "new_game_started")
        case .gameOver:
            .init(
                name: "game_over",
                params: [
                    "current_world": current_world_id(),
                    "kunai_count": ammo_in_inventory_for_weapon(UInt32(SPECIES_KUNAI_LAUNCHER), 0),
                    "rem223_count": ammo_in_inventory_for_weapon(UInt32(SPECIES_AR15), 0),
                    "cannonball_count": ammo_in_inventory_for_weapon(UInt32(SPECIES_CANNON), 0)
                ]
            )
        case .worldTransition(let source, let destination):
            if source != 0 {
                .init(
                    name: "world_transition",
                    params: [
                        "source": source,
                        "destination": destination
                    ]
                )
            } else {
                nil
            }
        }
    }
}

private struct AnalyticsEvent {
    let name: String
    let params: [String: Any]?
    
    init(name: String, params: [String: Any]? = nil) {
        self.name = name
        self.params = params
    }
}
